### ISSUE TITLE
Honor nested `beforeEvaluate` requests and others

### ISSUE THREAD

**Issue by bamboo** [gradle/gradle#5262]
### Current Behavior

Consider this code in `settings.gradle`:

```gradle
// settings.gradle
gradle.rootProject {
    beforeEvaluate {
        println "> Outer"
        beforeEvaluate {
            println "Inner"
        }
        print "< Outer"
    }
}
```

This now prints:

```text
> Outer
< Outer
```

And the `Inner` part is silently ignored.

### Expected Behavior

The nested requested should be honoured asynchronously and print:

```text
> Outer
< Outer
Inner
```
### Context

Hooks such as `beforeEvaluate` and `afterEvaluate` are used in a lot of ways to work around ordering problems, and it's easy to end up nesting two of them. Debugging such issues can be pretty hard.

See #3411 and #5200

**Comment by JLLeitschuh** [comment]
@bamboo That syntax isn't supported in the Gradle Kotlin DSL. What is `this` in the scope of the `gradle.rootProject`?

**Comment by JLLeitschuh** [comment]
Closing as closed in #5200
